### PR TITLE
[Feature branch] Implement data sources for MLflow artifact repositories

### PR DESCRIPTION
**Pull request by dbczumar**
## What changes are proposed in this pull request?

Implement data sources for MLflow artifact repositories, which will be used with MLflow Tracking.

This PR depends on https://github.com/mlflow/mlflow/pull/8046 (dataset source plugin system) and https://github.com/mlflow/mlflow/pull/8042 (dataset plugin system). The diff will shrink dramatically once those PRs are merged.

## How is this patch tested?

- [X] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/uri.py [hunk 1]
```diff
@@ -19,9 +19,16 @@
...
+def is_local_uri(uri, is_tracking_or_registry_uri=True):
+    """
+    Returns true if the specified URI is a local file path (/foo or file:/foo).
+
+    :param uri: The URI.
+    :param is_tracking_uri: Whether or not the specified URI is an MLflow Tracking or MLflow
+                            Model Registry URI. Examples of other URIs are MLflow artifact URIs,
+                            filesystem paths, etc.
+    """
+    if uri == "databricks" and is_tracking_or_registry_uri:
```
Comment (dbczumar) [inline mlflow/utils/uri.py:31]: If the `uri` isn't an MLflow Tracking or Model Registry URI, then "databricks" is perfectly valid

File: tests/data/test_artifact_dataset_sources.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    [
+        ("/tmp/path/to/my/local/file.txt", "local", "LocalArtifactDatasetSource"),
+        ("file:///tmp/path/to/my/local/directory", "local", "LocalArtifactDatasetSource"),
+        ("s3://mybucket/path/to/my/file.txt", "s3", "S3ArtifactDatasetSource"),
+        ("gs://mybucket/path/to/my/dir", "gs", "GCSArtifactDatasetSource"),
+        ("wasbs://user@host.blob.core.windows.net/dir", "wasbs", "AzureBlobArtifactDatasetSource"),
+        ("ftp://mysite.com/path/to/my/file.txt", "ftp", "FTPArtifactDatasetSource"),
+        ("sftp://mysite.com/path/to/my/dir", "sftp", "SFTPArtifactDatasetSource"),
+        ("hdfs://host_name:8020/hdfs/path/to/my/file.txt", "hdfs", "HdfsArtifactDatasetSource"),
+        ("viewfs://host_name:8020/path/to/my/dir", "viewfs", "HdfsArtifactDatasetSource"),
```
Comment (prithvikannan) [inline tests/data/test_artifact_dataset_sources.py:23]: Should this say `ViewFsArtifactDatasetSource`?

File: tests/data/test_artifact_dataset_sources.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+def test_local_downloads(tmp_path):
+    # Test string file paths
+    file_path = str(tmp_path / "myfile.txt")
+    with open(file_path, "w") as f:
+        f.write("text")
+
+    file_dataset_source = resolve_dataset_source(file_path)
+    assert file_dataset_source._get_source_type() == "local"
+    assert file_dataset_source.download() == file_path
```
Comment (prithvikannan) [inline tests/data/test_artifact_dataset_sources.py:98]: do we want to include a check here for the contexts of the file?

**Review by prithvikannan — APPROVED** [review]
Looks good! Think I caught a small typo and left a question

File: tests/data/test_artifact_dataset_sources.py (hunk 2 above)
Comment (dbczumar) [reply tests/data/test_artifact_dataset_sources.py:23]: `HdfsArtifactRepository` implements artifact operations for hdfs:/ and viewfs:/ locations, so `HdfsArtifactDatasetSource` is appropriate here.

File: tests/data/test_artifact_dataset_sources.py (hunk 3 above)
Comment (dbczumar) [reply tests/data/test_artifact_dataset_sources.py:98]: I think file contents are a better check, you're right. We may adjust the returned path later on :).

**Comment by dbczumar** [comment]
Merging!

**Pull request merged**